<div id="home-page-body" class="whole-page-bg">
    <div class="preloader">
      <div class="bar"></div>
    </div>
    <div id="page">
      <div id="heroWrapper" class="container-fluid">
        <div id="hero" class="container">
          <div class="container">
            <div class="row top-links">
              <div class="logo col-xs-3 col-sm-3 col-md-2 col-lg-2">
                <div class='zlab-logo-container'><img class="zlab-logo" alt="Logo" title="Zhang Lab" style="display:none"><span class="logo-desc"><a id="back-home-link" href="/#/">Back home</a></span></div>
              </div>
              <div class="social col-xs-9 col-sm-9 col-md-10 col-lg-10">
                <div class="social-links">
                  <a href="/#/about" class="top-left-link">About</a>
                  <a href="http://zlab.mit.edu/" class="top-left-link" style="display:none">Team</a>
                  <a href="mailto:guides-tech@googlegroups.com" class="top-left-link">Contact Us</a>
                  <a href="https://github.com/sanjanalab/GUIDES/issues" class="top-left-link">Report a Bug</a>
                </div>
              </div>
            </div>

            <div class="row main-content">
              <div id="about-page">
              	<div class="col-md-12">
                	<h3>What is Graphical User Interface for DNA Editing Screens (GUIDES)?</h3>
                	<p>The development of CRISPR (clustered regularly interspaced short palindromic repeat)-associated Cas9 nuclease has enabled new methods for interrogating gene function. Given the simple programmability of CRISPR nucleases through short RNA guides, genome-scale knockout libraries have emerged as popular tools for phenotypic screens [1]. Other screening applications may require libraries that target particular sets of genes, have deeper coverage with more guide RNAs per gene, or target specific cell types or tissues. GUIDES (Graphical User Interface for DNA Editing Screens) is a web application to design customized CRISPR knockout libraries as easily as possible without sacrificing control.</p>

              		<h3>How does it work?</h3>
              		<iframe width="560" height="315" src="https://www.youtube.com/embed/9_kASzIlLHI" frameborder="0" allowfullscreen></iframe>

                	<h3>Technical Specifications</h3>
                	<p>GUIDES provides a complete pipeline for the design of custom single-guide RNA (sgRNA) libraries for use in a wide variety of applications. Briefly, GUIDES first finds sgRNAs within the consensus coding sequences of a user-specific list of genes. To prioritize sgRNAs for each gene, it integrates tissue-specific exon expression from the GTEx Consortium [2], on-target scores for higher activity (Doench et al. and Microsoft Azimuth) and also off-target avoidance [3-4] to optimize guide selection. When possible, GUIDES also targets functional protein domains based on Pfam data [6], as mutagenesis of functional protein domains leads to a higher proportion of null mutations and enhanced severity of negative selection [6].</p>
                  <p>This website provides a ready-to-use version of GUIDES and, to promote further development and improvement of this tool, we also provide source codeÂ for GUIDES in our <a href="https://github.com/sanjanalab/GUIDES">GitHub repository</a>.</p>
                	<p>For further questions or to report problems, please contact us at <a href="mailto:guides-tech@googlegroups.com">guides-tech@googlegroups.com</a>.</p>
                  <h3>References</h3>
                  <ol>
                  <li>Shalem, O., Sanjana, N. E. &amp; Zhang, F. Nat. Rev. Genet. 16, 299-311 (2015).</li>
                  <li>The GTEx Consortium. Nat. Genet. 6, 580-585 (2013).</li>
                  <li>Doench, J. G., Fusi, N., Sullender, M., Hegde, M., Vaimberg, E. W., Donovan, K. F., Smith, I., Tothova, Z., Wilen, C., Orchard, R., Virgin, H. W., Listgarten, J. &amp; Root, D. E. Nat. Biotechnol. 2, 184-191 (2016).</li>
                  <li>Doench, J. G., Hartenian, E., Graham, D. B., Tothova, Z., Hegde, M., Smith, I., Sullender, M., Ebert, B. L., Xavier, R. J. &amp; Root, D. E. Nat. Biotechnol. 12, 1262-1267 (2014).</li>
                  <li>Finn, R. D., Coggill, P., Eberhardt, R. Y., Eddy, S. R., Mistry, J., Mitchell, A. L., Potter, S. C., Punta, M., Qureshi, M., Sangrador-Vegas, A., Salazar, G. A., Tate, J. &amp; Bateman, A. Nucleic Acids Res. 44, D279-D285 (2015).</li>
                  <li>Shi, J., Wang, E., Milazzo, J. P., Wang, Z., Kinney, J. B. &amp; Vakoc, C. R. Nat. Biotechnol. 33, 661-667 (2015).</li>
              	</div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
